### ISSUE TITLE
When using pre-trained weights for training, an error is reported

### ISSUE THREAD

**Issue by meiwu5** [KAIST-VICLab/FMA-Net#15]
When using pre-trained weights for training, an error is reported  “optimizer_D_state_dict  miss”，Please tell me how to solve this problemPlease tell me how to solve this problem
![image](https://github.com/KAIST-VICLab/FMA-Net/assets/142979824/9048b142-9b8a-4771-9992-9e840b5ac466)

**Comment by meiwu5** [comment]
Hello, I use your pre-trained weights and fine-tune with my own data, but an error occurs when loading. I look forward to your reply.

**Comment by GeunhyukYouk** [comment]
Hello, I apologize for the late response.

If you would like to use the pre-trained FMA-Net, please first download the weights from this [link](https://www.dropbox.com/scl/fo/4392nxna1wptrw06ktv6r/AIyy20JrXK_9CMcXHUQY7Ko?rlkey=n4hhgl7p2c63y3l6lkpqlthi0&e=1&st=mnmmvm9y&dl=0).
Then, move the two downloaded weights to './results/FMA-Net/model_stage2', and run the code. Alternatively, you can set the [save_dir](https://github.com/KAIST-VICLab/FMA-Net/blob/a7a595d88a64014022b8f04eaec0622e861b1a31/experiment.cfg#L6) in [experiment.cfg](https://github.com/KAIST-VICLab/FMA-Net/blob/main/experiment.cfg) to your desired path.

**Comment by meiwu5** [comment]
When I use the pre-trained model (using the original weights and fine-tuning my own dataset), I go to line 32 in the main function, " if config.stage == 2: trainer.load_best_stage1_model() "
, and an error occurs when loading the first stage model, saying that ‘optimizer_D_state_dict’ is not found. The data of the original model is downloaded from https://www.dropbox.com/scl/fo/4392nxna1wptrw06ktv6r/AIyy20JrXK_9CMcXHUQY7Ko?rlkey=n4hhgl7p2c63y3l6lkpqlthi0&st=mnmmvm9y&dl=0. 

I read your answer and put the weights in ‘./results/FMA-Net/model_stage2’, but I debugged and found that this is not loaded at runtime, and it will load trainer.load_best_stage1_model() first.

**Comment by GeunhyukYouk** [comment]
This happened because you don't have a pre-trained Stage 1 model. 

If you simply want to **fine-tune the FMA-Net on your dataset**, this is not a problem and can be ignored. You can simply delete lines 31 and 32 in main.py. 

Also, don't forget to appropriately adjust the loss weights during fine-tuning.

**Comment by meiwu5** [comment]
Sorry to bother you again

Background: I hope to use the pre-train model you provided to do fine-tuning based on my own dataset.

I did what you said, commented out lines 31 and 32 in main, set config.finetuning to true, and ran the code in main, where last_epoch = trainer.load_checkpoint() will load "./results/FMA-Net/model_stage2/model_D_latest.pt". I renamed the pre-trained model "model_D_best ; model_D_best." to "model_D_latest.pt; model_R_latest.pt” (downloaded from https://www.dropbox.com/scl/fo/4392nxna1wptrw06ktv6r/AIyy20JrXK_9CMcXHUQY7Ko?rlkey=n4hhgl7p2c63y3l6lkpqlthi0&st=mnmmvm9y&dl=0.)

When the code is loading the model weight, it reports an error that ‘optimizer_D_state_dict
’ is not found,

I debugged and looked at the key-value pairs of D_state_dict, and there is no ‘optimizer_D_state_dict’. 

Is it that the pre-train model provided by github does not have ‘optimizer_D_state_dict’, or is there something wrong with my configuration?

**Comment by KAIST-VICLab** [comment]
Hello,

When I uploaded the pretrained model, I removed unnecessary optimizers for inference. I apologize for any inconvenience caused by not considering this scenario. However, since you're fine-tuning on a new dataset, it's okay to train with a new optimizer.

Could you please try changing line 37 in main.py to `trainer.load_best_model()` and run it again? Let me know if you encounter any further issues!

**Comment by GeunhyukYouk** [comment]
Hello,

When I uploaded the pretrained model, I removed unnecessary optimizers for inference. I apologize for any inconvenience caused by not considering this scenario. However, since you're fine-tuning on a new dataset, it's okay to train with a new optimizer.

Could you please try changing line 37 in main.py to `trainer.load_best_model()` and run it again? Let me know if you encounter any further issues!

**Comment by GeunhyukYouk** [comment]
Also, I'm curious if you have prepared the pseudo-GT flow generated by RAFT for the flow training process (corresponding to $`l_1(f^Y, f^Y_{RAFT})`$ of Eq. 8 in the [paper](https://openaccess.thecvf.com/content/CVPR2024/html/Youk_FMA-Net_Flow-Guided_Dynamic_Filtering_and_Iterative_Feature_Refinement_with_Multi-Attention_CVPR_2024_paper.html)). If not, you can generate it using [generate_flow.py](https://github.com/KAIST-VICLab/FMA-Net/blob/main/preprocessing/generate_flow.py), or you can choose not to utilize it at the cost of a slight performance loss.
